controllers/clustergroupupgrade_controller.go

```diff
@@ -327,9 +327,23 @@ func (r *ClusterGroupUpgradeReconciler) Reconcile(ctx context.Context, req ctrl.
 						nextReconcile = ctrl.Result{RequeueAfter: requeueAfter}
 					}
 
-					// Check for batch timeout.
-					batchTimeout := time.Duration(clusterGroupUpgrade.Spec.RemediationStrategy.Timeout/len(clusterGroupUpgrade.Status.RemediationPlan)) * time.Minute
-					if !clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt.IsZero() && time.Since(clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt.Time) > batchTimeout {
+					// The remaining time will be the total timeout subtract the elapsed time spent on both the batch and cgu
+					// It is important to include the current batch here so that we don't let the entire timeout be consumed by a single batch that gets stuck
+					remainingTime := float64(clusterGroupUpgrade.Spec.RemediationStrategy.Timeout) - clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt.Time.Sub(clusterGroupUpgrade.Status.Status.StartedAt.Time).Seconds()
```

Comment:
better to check for nil value before clusterGroupUpgrade.Status.Status.CurrentBatchStartedAt.Time

Reply:
I can reorganize the if checks to make sure this is checked for zero before it is used